src/relay/ir/doc.h

```diff
@@ -0,0 +1,436 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file doc.h
+ * \brief A pretty printer DSL for constructing (Doc) and formatting (RDoc) documents.
+ *        It is based heavily on Philip Wadler's "A prettier printer."
+ *        See https://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf
+ *        for more details.
+ *
+ * Since the original paper uses call by value for efficiency, everything doc function is maximally lazy.
+ * You can probably yank speed by doing strict analysis and removing some Lazy (if this is bottleneck).
+ */
+#ifndef TVM_RELAY_IR_DOC_H_
```

Comment:
This code is quite functional and to be honest I do not know what was happening in here. Is it possible to add comments on what is going on? 

Ideally, if we can get a PrettyPrinter with "normal code" patterns like Visitor, that is preferred since more developer will be able to understand what is going on

Reply:
I had add some comments since making the pr, and IMO it is easier to read the original paper then to read this file. There is also no need to change (nor has it been changed) the core logic of the file since it is a mechanical translation (given time I can write a program to do it), and at later stage we can look back and decide what to do? (since there is still tons of work to be ported).

Reply:
Sorry, as we move upstream we do want to keep hold higher standard on the code being merged, so please polish as much as we can.

Reply:
In the case of public header file that is getting used, we need to document the data structure and what does each field mean